api/envoy/config/core/v3/grpc_service.proto

```diff
@@ -55,6 +56,16 @@ message GrpcService {
     // This limit is applied to individual messages in the streaming response and not the total size of streaming response.
     // Defaults to 0, which means unlimited.
     google.protobuf.UInt32Value max_receive_message_length = 4;
+
+    // [#not-implemented-hide:]
+    // If true, x-envoy-internal header will be added to Envoy-gRPC request header.
+    // Default to true.
+    google.protobuf.BoolValue add_internal_header = 5;
+
+    // [#not-implemented-hide:]
+    // If true, x-forwarded-for header will be added to Envoy-gRPC request header.
```

Comment:
I realize that we're preserving backwards compatible behavior here by making the default true, but I think XFF for async client makes no sense. Async requests are sidecalls, we're not forwarding to the service on behalf of some client, we're making an RPC to some service mid-way through request processing. For gRPC in particular, I think it's a confusion to include XFF by default in the "gRPC metadata" - @markdroth.

Reply:
Make sense. xff is to provide a client IP address to the server which is not applicable to sidecar case. 

We probably should use runtime guard approach to make xff removal default. I removed this config from this PR.

PTAL, Thanks

Reply:
I think this makes sense based on my understanding of the original intent from https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_conn_man/headers.html#x-envoy-internal, although, again it's a bit weird to be attaching something like `x-envoy-internal` to gRPC - the destination service should know that any call to it is coming from a set of internal trusted clients.

@mattklein123 for a second opinion on this.

Reply:
Friendly ping @mattklein123 's opinion.

I think the fundamental question here is : 
Whether (1) we want to remove `x-enovy-internal` from envoy-gRPC path by default. OR (2) we want to keep `x-envoy-internal`  but enable specific gRPC client's capability of removing it.

From strictly backwards-compatibility perspective, (2) is better.  But if we are in agreement with end-state of removing this header from envoy-gRPC, (1) is good and its impact can be mitigated by runtime guard.

Reply:
I went ahead with (1) with runtime guard. PTAL

Reply:
(reactivate this thread)

Based on my understanding of @mattklein123's [comment here](https://github.com/envoyproxy/envoy/pull/34857#pullrequestreview-2135934179), the suggested approach is adding config options for `x-envoy-internal` and `x-forward-for` respectively, which is this PR's approach.

@htuch  WDYT? If you also agree with it, I am going to proceed with this PR and add back config option for xff

Thanks!

Reply:
Yes, config guarding makes sense, thanks.

Reply:
I have update the PR to include both API and implementation. Also [PR description](https://github.com/envoyproxy/envoy/pull/34179#issue-2298360332) has details on the design logic .

PTAL, Thanks!